### ISSUE TITLE
HttpClient - No supported cipher suites leads to stuck requests.

### ISSUE THREAD

**Issue by veselinandreev** [eclipse/jetty.project#377]
It appears that whenever you attempt to connect to an end-point which enforces specific set of cipher suites, none of which supported by the client (intentionally or otherwise restricted by JVM, i.e. 256bit ones), the request remains in the queue and never gets aborted/cancelled. Below exception has been swallowed by `SelectorManager.connectionOpened(Connection connection)`, thus never propagated further up the chain and so client is still waiting on its connection:
```
org.eclipse.jetty.io.RuntimeIOException: javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)
	at org.eclipse.jetty.io.ssl.SslConnection.onOpen(SslConnection.java:150) ~[jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager.connectionOpened(SelectorManager.java:310) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.createEndPoint(SelectorManager.java:747) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.processConnect(SelectorManager.java:681) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.processKey(SelectorManager.java:644) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.select(SelectorManager.java:611) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.io.SelectorManager$ManagedSelector.run(SelectorManager.java:549) [jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	at org.eclipse.jetty.util.thread.NonBlockingThread.run(NonBlockingThread.java:52) [jetty-util-9.2.14.v20151106.jar:9.2.14.v20151106]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [?:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [?:1.7.0_79]
	at java.lang.Thread.run(Thread.java:745) [?:1.7.0_79]
Caused by: javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)
	at sun.security.ssl.Handshaker.activate(Handshaker.java:470) ~[?:1.7.0_79]
	at sun.security.ssl.SSLEngineImpl.kickstartHandshake(SSLEngineImpl.java:714) ~[?:1.7.0_79]
	at sun.security.ssl.SSLEngineImpl.beginHandshake(SSLEngineImpl.java:740) ~[?:1.7.0_79]
	at org.eclipse.jetty.io.ssl.SslConnection.onOpen(SslConnection.java:143) ~[jetty-io-9.2.14.v20151106.jar:9.2.14.v20151106]
	... 10 more
```

By the point `SelectorManager.connectionOpened(Connection connection)` is called, the `ConnectTimeout` task has been cancelled, so the particular request (and any future to the same destination) never gets out of the queue. In case however a 'Total timeout' (specified through `Request.timeout(long timeout, TimeUnit unit)`) has been specified, the request will be aborted once the time have elapsed, although with a different error ('Total timeout elapsed').

Given that `SSLHandshakeException` (and any other `SSLException` derivative) is an IOException should the declaration of `Connection.onOpen()` and `SelectorManager.connectionOpened(Connection connection)` be changed to allow IOException's, thus propagating the I/O error through `SelectorManager.ManagedSelector.createEndPoint(SocketChannel channel, SelectionKey selectionKey)` to `
SelectorManager.ManagedSelector.processConnect(SelectionKey key, Connect connect)` where it would be dealt accordingly ?

Kind Regards,
Ves.

**Comment by sbordet** [comment]
Veselin,

can you please share your `SslContextFactory` configuration that you used to obtain this error ?
I want to write a test case for this.

Thanks !

**Comment by veselinandreev** [comment]
Hi Simone, 
  below should be sufficient to trigger the issue (assuming you have not enabled 256b ciphers):
```
factorySSLContext.setIncludeCipherSuites("TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384");
```

So in effect, the client has no available cipher suite to use. My original report implied its due to different ciphers on both sides, but that's another issue which I've mixed with this one, apologies.

That said, if you use the below configuration, client would get EOF rather than handshake exception.
- at the server side:
```
factorySSLContext.setIncludeCipherSuites("TLS_RSA_WITH_AES_128_CBC_SHA");
```
- At the client side:
```
factorySSLContext.setIncludeCipherSuites("TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA");
```

In case of handshake issues, I would expect to get something along the lines:
```
javax.net.ssl.SSLHandshakeException: Received fatal alert: handshake_failure
```

Kind Regards,
Ves.

**Comment by sbordet** [comment]
Fixed in 9.2.x and subsequent.

The actual issue came from a misconfiguration of the client, where `SslContextFactory` was configured with ciphers that were not good for the TLS protocol chosen (e.g. TLSv1.1 with TLSv1.2 ciphers), and I was able to reproduce the issue.

Rather than modifying the signature of onOpen(), I just rethrew the exception internally so that it could be handled properly.

Note that this issue only affected 9.2.x since in 9.3.x we don't call `beginHandshake()` anymore from `onOpen()`, but the fix was nevertheless applies to 9.3.x and subsequent.

Thanks !

**Comment by veselinandreev** [comment]
Thanks, just worth mentioning that the TLS protocol doesn't really play a role here, its more to do with the ciphers that has been allowed at the client side and whether same are enabled in the JVM. Configuring both client & server on TLSv1.1, but have the client ciphers restricted to one that is intentionally disabled would result in same issue.

With regards to the server side issue, does this fix resolve the problem there as well ? i.e. would the client receive a handshake failure or EOF ?

**Comment by sbordet** [comment]
Veselin,

this issue was about `HttpClient` being stuck at sending requests in case the TLS protocol and the ciphers don't play well together, and it's solved.

On server side there was never a problem.

The application would receive a failure. Which type of failure is often very difficult to tell and depends on the underlying transport. There is a natural race condition for the client writing and the client reading; depending on which of the two fails first, the client may get a different error.

Note however that this is a configuration issue; such configured `HttpClient` won't work unless its TLS configuration is changed, which typically needs stop and restart.

Bottom line is to not rely on the type of failure, and configure TLS properly ahead of time.

**Comment by veselinandreev** [comment]
Hi Simone,
  thanks for fixing the issue within the `HttpClient`! Much appreciated!

As for the server side, whilst I agree `HttpClient` needs to be configured correctly, the fact that the server is dropping the connection rather than reporting the handshake failure is of concern. 

Connecting to an external service, that is powered by Jetty, and having connection drops instead of handshake failures can lead to several hours of investigation and if the other party is not willing to put much of resources to help you, this can become rather long night of debugging.

Now I'm no expect on the TLS protocol, but according to RFC 5246, an error alert is to be triggered at least.

7.2.2.  Error Alerts:
```
 Error handling in the TLS Handshake protocol is very simple.  When an
   error is detected, the detecting party sends a message to the other
   party.  Upon transmission or receipt of a fatal alert message, both
   parties immediately close the connection.  Servers and clients MUST
   forget any session-identifiers, keys, and secrets associated with a
   failed connection.  Thus, any connection terminated with a fatal
   alert MUST NOT be resumed.

Whenever an implementation encounters a condition which is defined as
   a fatal alert, it MUST send the appropriate alert prior to closing
   the connection.  For all errors where an alert level is not
   explicitly specified, the sending party MAY determine at its
   discretion whether to treat this as a fatal error or not.  If the
   implementation chooses to send an alert but intends to close the
   connection immediately afterwards, it MUST send that alert at the
   fatal alert level.
```

Now consider a case where a jetty powered web service is enforcing 256b ciphers, so a client has to support them, but JVM by default have them disabled, so it may not come that obvious to whoever debugging this what exactly is causing the connection drops.

**Comment by sbordet** [comment]
Veselin,

If the client is misconfigured, the client itself will blow up before trying to send any bytes to the server.
There is no network communication between the client and the server apart the initial TCP SYNs to establish the connection.

If the server is misconfigured, unfortunately there is no way for the server to let the client know what kind of error it got, because there are no Java APIs that allow to customize the TLS Alert message with a specific error code. The client sees a generic TLS Alert (probably just a close) or just the connection closed at TCP level. Either case gives the client zero information.

The only option we would have would be to write ourselves the TLS bytes (reimplementing ourselves the TLS protocol without using what provided by the JDK). Doable, but we have other fishes to fry ;)

If you have time and you want to try it out, you're welcome to open a new issue and let's discuss it over there.
